### PR TITLE
[SPARK-13473][SQL] Don't push predicate through project with nondeterministic field(s)

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

Predicates shouldn't be pushed through project with nondeterministic field(s).

See https://github.com/graphframes/graphframes/pull/23 and SPARK-13473 for more details.

This PR targets master, branch-1.6, and branch-1.5.
## How was this patch tested?

A test case is added in `FilterPushdownSuite`. It constructs a query plan where a filter is over a project with a nondeterministic field. Optimized query plan shouldn't change in this case.

### CODE REVIEW

**Comment by liancheng** [comment]
cc @mengxr

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala [hunk 1]
```diff
@@ -156,6 +156,17 @@ class FilterPushdownSuite extends PlanTest {
     comparePlans(optimized, originalQuery)
   }
 
+  test("nondeterministic: can't push down filter through project with nondeterministic field") {
+    val originalQuery = testRelation
+        .select(Rand(10).as('rand), 'a)
```
Comment (mengxr) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala]: minor: indentation

**Comment by marmbrus** [comment]
LGTM

**Comment by liancheng** [comment]
We should also revise nondeterminism handling in `PushPredicateThroughGenerate`, `PushPredicateThroughAggregate`, and `PushPredicateThroughJoin`. But they can be added in follow-up PRs.

**Comment by liancheng** [comment]
test this please

**Comment by yhuai** [comment]
test this please

**Comment by liancheng** [comment]
We can probably further simplify `PushPredicateThroughProject`. Before this PR, we don't push a predicate if it refers to any non-deterministic field(s). However, as what this PR fixes, we shouldn't push a predicate through any project that has non-deterministic field(s). Another case worth noting is that a predicate containing non-deterministic expression(s) but not referring to any non-deterministic field(s) is OK to be pushed down. For example, it's OK to push down the following filter predicate:

```scala
// from:
sqlContext.range(3).select('id as 'a, 'id 'as 'b).filter(rand(42) > 0.5)

// to:
sqlContext.range(3).filter(rand(42) > 0.5).select('id as 'a, 'id 'as 'b)
```

This means that we can push down a filter predicate through a project *if and only if* all fields of the project are deterministic. That's why those two test cases are considered outdated and removed.

cc @cloud-fan 

(To be safe, I won't do the above update in this PR since it also targets to 1.6 and 1.5.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -804,7 +804,9 @@ object SimplifyFilters extends Rule[LogicalPlan] {
  */
 object PushPredicateThroughProject extends Rule[LogicalPlan] with PredicateHelper {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case filter @ Filter(condition, project @ Project(fields, grandChild)) =>
+    case filter @ Filter(condition, project @ Project(fields, grandChild))
+      if fields.forall(_.deterministic) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: how about we add some comments to explain why we can't push down filter through project with non-deterministic fields? e.g. number of input row is also an implicit input for non-deterministic expressions, push down filter will break it.

**Comment by cloud-fan** [comment]
LGTM except one comment

**Comment by liancheng** [comment]
Thanks for the review, comment added.

**Comment by cloud-fan** [comment]
The last commit is adding comments, so it's safe to merge after style check passed.
Thanks, merged into master, branch-1.6 and branch-1.5!

**Comment by wecharyu** [comment]
Hi @liancheng, @mengxr, @cloud-fan,

I'm trying to understand why pushing down filters with nondeterministic fields is considered a bug. How would the different nondeterministic results impact the query?

For instance, other engines like Hive do push down filters in these cases. Could this change lead to performance regressions in our queries?

**Pull request closed without merging**